lib/cog/adapters/hipchat/api.ex

```diff
@@ -1,153 +1,219 @@
 defmodule Cog.Adapters.HipChat.API do
-  alias Cog.Adapters.HipChat
+  use GenServer
   require Logger
 
-  use HTTPoison.Base
-
-  defstruct token: nil
-
-  @api_base "https://api.hipchat.com/v2"
+  @base_uri "https://api.hipchat.com/v2"
   @timeout 15000 # 15 seconds
 
-  def send_message(%{"id" => room_id}, message) do
-    url = "/room/#{room_id}/message"
-    body = Poison.encode!(%{message: message})
-    {:ok, response} = post(url, body)
-    response.body
+  def start_link(config) do
+    GenServer.start_link(__MODULE__, config, name: __MODULE__)
   end
-  def send_message(%{"direct" => user_id}, message) do
-    url = "/user/#{user_id}/message"
-    body = Poison.encode!(%{message: message})
-    {:ok, response} = post(url, body)
-    response.body
+
+  def send_message(room, message) do
+    GenServer.call(__MODULE__, {:send_message, room, message})
   end
 
-  def lookup(subject, id, expands \\ nil) do
-    get!("/#{subject}/#{id}" <> expands_list(expands))
+  def lookup_room("@" <> handle) do
+    user = lookup_user(name: "@" <> handle)
+    lookup_direct_room(user_id: user.id)
   end
 
-  def lookup_direct_room(user_id: user_id),
-    do: {:ok, %{"direct" => user_id}}
+  def lookup_room("#" <> room) do
+    lookup_room(name: room)
+  end
 
-  defp expands_list(nil), do: ""
-  defp expands_list(expands) do
-    "?expand=" <> Enum.join(expands, ",")
+  def lookup_room(options) do
+    GenServer.call(__MODULE__, {:lookup_room, options})
   end
 
-  # TODO: cache looking up the users in the room
-  def lookup_room_user(room_name, user_name) do
-    room = lookup("room", room_name).body
-    members = get!("/user").body["items"]
+  def lookup_direct_room(user_id: user_id) do
+    {:ok, %{"direct" => user_id}}
+  end
 
-    user = Enum.find(members, fn(user) ->
-      Map.get(user, "name") == user_name
-    end)
+  def lookup_room_user(room_name, user_name) do
+    with {:ok, room} <- lookup_room(name: room_name),
+         {:ok, users} <- lookup_users() do
+      user = Enum.find(users, &match?(%{name: ^user_name}, &1))
```

Comment:
Not sure, this was just pulled from the old codebase. Let me try that out before I merge this.